src/node.js

```diff
@@ -390,6 +391,61 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var pendingRejections = [];
```

Comment:
Yes splicing is slow too (can implement custom splice if you need), but nothing compared to using maps and iterator protocols.